src/components/Form.spec.jsx

```diff
@@ -114,3 +115,38 @@ test('Start form with object reference query param', async () => {
   const requestBody = await startSubmissionRequest.json();
   expect(requestBody.initialDataReference).toBe('foo');
 });
+
+// Regression test for https://github.com/open-formulieren/open-forms/issues/4918
+test.each([
+  {
+    introductionPageContent: '',
+    buttonText: 'Login with DigiD',
+    expectedUrl:
+      'http://mock-digid.nl/login?next=http%3A%2F%2Flocalhost%2F%3F_start%3D1%26initial_data_reference%3Dfoo',
+  },
+  {
+    introductionPageContent: 'foo',
+    buttonText: 'Continue',
+    expectedUrl: '/startpagina?initial_data_reference=foo',
+  },
+])(
+  'Redirect to start page or introduction page should preserve initial_data_reference param',
+  async ({introductionPageContent, buttonText, expectedUrl}) => {
+    mswServer.use(mockAnalyticsToolConfigGet(), mockSubmissionPost(), mockSubmissionStepGet());
+
+    render(
+      <Wrapper
+        form={buildForm({
+          loginOptions: [{identifier: 'digid', label: 'DigiD', url: 'http://mock-digid.nl/login'}],
+          introductionPageContent: introductionPageContent,
+        })}
+        initialEntry={`/?initial_data_reference=foo`}
+      />
+    );
+
+    await waitFor(() => {
+      const loginLink = screen.getByRole('link', {name: buttonText});
+      expect(loginLink).toHaveAttribute('href', expectedUrl);
+    });
```

Comment:
you can use the `findBy*` queries, which are async (and basically put the  `waitFor` inside the library
```suggestion
    const loginLink = await screen.findByRole('link', {name: buttonText});
    expect(loginLink).toHaveAttribute('href', expectedUrl);
```